include/seqan3/alphabet/concept.hpp

```diff
@@ -16,22 +16,936 @@
 #include <seqan3/core/metafunction/function.hpp>
-#include <seqan3/alphabet/adaptation/pre.hpp>
+#include <seqan3/alphabet/adaptation/char.hpp>
+#include <seqan3/alphabet/adaptation/uint.hpp>
 #include <seqan3/alphabet/concept_pre.hpp>
 #include <seqan3/alphabet/detail/member_exposure.hpp>
+#include <seqan3/alphabet/exception.hpp>
 #include <seqan3/std/concepts>
 
+// ============================================================================
+// to_rank()
+// ============================================================================
+
+namespace seqan3::detail::adl::only
+{
+
+//!\brief Call `to_rank()` on the argument, but only consider ADL.
+//!\ingroup alphabet
+template <typename alph_t>
+//!\cond
+    requires requires { to_rank(std::declval<alph_t>()); }
+//!\endcond
+constexpr decltype(auto) to_rank_(alph_t && a) noexcept(noexcept(to_rank(std::declval<alph_t>())))
+{
+    return to_rank(std::forward<alph_t>(a));
+}
+
+} // namespace seqan3::detail::adl::only
+
+namespace seqan3::detail
+{
+
+// Auxiliary concepts
+//!\cond
+template <typename alph_t>
+SEQAN3_CONCEPT has_to_rank_adl_overload = requires { { adl::only::to_rank_(std::declval<alph_t>()) }; };
+
+template <typename alph_t>
+SEQAN3_CONCEPT has_to_rank_adaptation_overload = requires { { ::seqan3::adaptation::to_rank(std::declval<alph_t>()) }; };
+
+template <typename alph_t>
+SEQAN3_CONCEPT has_to_rank_member = requires { { std::declval<alph_t>().to_rank() }; };
+//!\endcond
+
+//!\brief Functor definition for seqan3::to_rank.
+struct to_rank_fn
+{
+    //!\brief Operator definition.
+    template <typename alph_t>
+    //!\cond
+        requires has_to_rank_adl_overload<alph_t>        ||
+                 has_to_rank_adaptation_overload<alph_t> ||
+                 has_to_rank_member<alph_t>
+    //!\endcond
+    constexpr decltype(auto) operator()(alph_t const a) const noexcept
+    {
+        if constexpr (has_to_rank_adl_overload<alph_t>)
+        {
+            static_assert(noexcept(adl::only::to_rank_(a)),
+                          "Only overloads that are marked noexcept are picked up by seqan3::to_rank().");
+            return adl::only::to_rank_(a);
+        }
+        else if constexpr (has_to_rank_adaptation_overload<alph_t>)
+        {
+            static_assert(noexcept(::seqan3::adaptation::to_rank(a)),
+                          "Only overloads that are marked noexcept are picked up by seqan3::to_rank().");
+            return ::seqan3::adaptation::to_rank(a);
+        }
+        else
+        {
+            static_assert(noexcept(a.to_rank()),
+                          "Only members that are marked noexcept are picked up by seqan3::to_rank().");
+            return a.to_rank();
+        }
+    }
+};
+
+} // namespace seqan3::detail
+
+namespace seqan3
+{
+
+/*!\name Function objects
+ * \{
+ */
+
+/*!\brief Return the rank representation of an (Semi-)Alphabet object.
+ * \param alph The target object; its type must model seqan3::Alphabet.
```

Comment:
must model (Semi-)Alphabet
Is this enforced somewhere? It is rather must have either a to_rank member, etc...?

Reply:
> It is rather must have either a to_rank member, etc...?

Yes. This is a little tricky, because when designing an alphabet, you need to provide an implementation → it is not the functor that requires the concept, but the other way around. However for 99% of the users it is "a functor that works on all types that model the concept".

I can change the documentation to be strictly correct, but it will be slightly less useful to our users.

Reply:
Hm.. The "strictly correct" documentation might be too verbose. But "must model alphabet" is not correct as this function is not requiring it but, as you said, helps types to model Alphabet. So this is confusing with how we document most other stuff. :(

Reply:
But I changed it, or not?